### PR TITLE
feat(extractor): fromPublished to extract from published packages

### LINKED ISSUE #68: Allow extracting docs from published packages on NPM

### LINKED ISSUE #69: Add real packages to test suite integration tests
Depends on https://github.com/tydoc/tydoc/issues/68

### PR DESCRIPTION
**Pull request by williamluke4**
closes #68 
closes #69

### CODE REVIEW

**Comment by jasonkuhrt** [comment]
I would rather not add all these files to the repo. Maybe we can snapshot tests against npm packages. Since npm packages are immutable and the registry is designed for constant usage it should be pretty reliable.

**Comment by williamluke4** [comment]
> I would rather not add all these files to the repo. Maybe we can snapshot tests against npm packages. Since npm packages are immutable and the registry is designed for constant usage it should be pretty reliable.

@jasonkuhrt I've changed it to dynamically pull from npm. You just need to put the name of the npm package [here](https://github.com/tydoc/tydoc/blob/feat/package-tests/packages/extractor/test/packages.test.ts#L13) and it will pull the latest version.

**Comment by williamluke4** [comment]
@jasonkuhrt I have also moved my work to [fromPublished.ts](https://github.com/tydoc/tydoc/blob/feat/package-tests/packages/extractor/src/api/extractor/fromPublished.ts#L83) which is now working. You can run `yarn jest ./test/packages.test.ts` to test it out. 

There is a lot of duplicated code currently but we can sort that out later

**Comment by jasonkuhrt** [comment]
> but we can sort that out later

Yeah, and #65 is going to have an impact too I suspect.

**Comment by jasonkuhrt** [comment]
I will probably review this Monday night EST.

File: packages/extractor/package.json [hunk 1]
```diff
@@ -34,37 +34,41 @@
...
+    "@types/decompress": "^4",
     "@types/dedent": "^0.7.0",
-    "@types/jest": "25.1.3",
+    "@types/jest": "26.0.15",
     "@types/json5": "^0.0.30",
-    "@types/node": "13.7.6",
-    "@types/prettier": "1.19.0",
+    "@types/node": "14.14.10",
+    "@types/prettier": "2.1.5",
+    "axios": "^0.21.0",
```
Comment (jasonkuhrt) [inline packages/extractor/package.json:60]: Used to be my goto as well. But what about https://github.com/sindresorhus/got?

**Review by jasonkuhrt — CHANGES_REQUESTED** [review]
More to come, just a start

File: packages/extractor/package.json (hunk 1 above)
Comment (williamluke4) [reply packages/extractor/package.json:60]: Will have a look. That guy is a god

File: packages/extractor/package.json [hunk 2]
```diff
@@ -34,37 +34,41 @@
...
+    "jest-validate": "26.6.2",
     "jest-watch-suspend": "1.1.2",
-    "jest-watch-toggle-config": "1.0.2",
-    "jest-watch-typeahead": "0.4.2",
-    "ts-jest": "25.2.1",
-    "ts-node": "8.6.2",
-    "ts-node-dev": "1.0.0-pre.44",
+    "jest-watch-toggle-config": "2.0.1",
+    "jest-watch-typeahead": "0.6.1",
+    "tempy": "^1.0.0",
```
[the remark this answers is not in the document]
Comment (williamluke4) [reply packages/extractor/package.json:68]: Had no idea it had one

File: packages/extractor/src/api/extractor/extract.ts [hunk 3]
```diff
@@ -534,4 +534,4 @@ function getTSDoc(manager: Doc.Manager, t: tsm.Type): Doc.TSDocFrag {
   return {
     tsdoc: docFragTSDoc,
   }
-}
+}
```
[the remark this answers is not in the document]
Comment (williamluke4) [reply packages/extractor/src/api/extractor/extract.ts:537]: Possibly prettier.. Will chech

File: packages/extractor/package.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (jasonkuhrt) [reply packages/extractor/package.json:68]: Yeah it's [only the last few releases](https://github.com/szwacz/fs-jetpack/issues/95)!

File: packages/extractor/test/packages.test.ts [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+  const tarPath = await downloadTarball(tarUrl)
+  await extract(tarPath, dir)
+}
+
+async function getTarUrl(packageName: string) {
+  let tarUrl: string
+  try {
+    const { body } = await got.get('https://registry.npmjs.org/' + packageName, { responseType: 'json' })
+    // @ts-ignore
+    const latestVersion = body['dist-tags']['latest']
```
Comment (jasonkuhrt) [inline packages/extractor/test/packages.test.ts:55]: We want to pick an immutable version.

**Comment by jasonkuhrt** [comment]
Main things left:

- sourceDir concept doesn't work when extracting from packages
- EDD paths should be relative, not absolute, at least for EDD from packages
- need a layout calculator suited to packages `getLayoutForPackage(path)`

**Comment by jasonkuhrt** [comment]
meh good enough

**Pull request merged**